File: src/modules/cliente/carrinho.ts
"use server"

import { revalidatePath } from "next/cache"

import { criarClienteDoServidor } from "@/lib/supabase/servidor"
import type { Enums } from "@/types/banco"
import { exigirUsuario, obterContexto } from "@/modules/auth/sessao"

/**
 * O carrinho do cliente, na web.
 *
 * Mora no servidor, como no aplicativo: e da tabela `carts` que
 * `public.fechar_pedido` le para recalcular os precos. Um carrinho so no
 * navegador obrigaria a mandar quanto custa - exatamente o que o banco nao
 * aceita.
 *
 * Consequencia boa: quem monta o pedido no computador e fecha no celular
 * encontra o carrinho como deixou.
 */

export type ResultadoDaAcao = { ok: true; id?: string } | { ok: false; erro: string }

function recarregar() {
  revalidatePath("/carrinho")
  revalidatePath("/", "layout")
}

/**
 * Abre (ou reabre) o carrinho de um estabelecimento.
 *
 * Trocar de restaurante apaga o carrinho anterior: um pedido pertence a um
 * estabelecimento so, e e o indice `carts_open_per_restaurant_idx` que garante
 * isso. Quem avisa o cliente antes de descartar e a tela.
 */
async function abrirCarrinho(restauranteId: string) {
  const contexto = await exigirUsuario("/carrinho")
  const supabase = await criarClienteDoServidor()

  const { data: existente } = await supabase
    .from("carts")
    .select("id, restaurant_id")
    .eq("user_id", contexto.id)
    .maybeSingle()

  if (existente?.restaurant_id === restauranteId) return existente.id
  if (existente) await supabase.from("carts").delete().eq("id", existente.id)

  const { data: novo, error } = await supabase
    .from("carts")
    .insert({ user_id: contexto.id, restaurant_id: restauranteId })
    .select("id")
    .single()

  if (error || !novo) throw new Error(error?.message ?? "Não consegui abrir o carrinho.")
  return novo.id
}

export async function adicionarAoCarrinho(dados: {
  restauranteId: string
  produtoId: string
  quantidade: number
  adicionais: string[]
  observacao?: string
}): Promise<ResultadoDaAcao> {
  try {
    const carrinhoId = await abrirCarrinho(dados.restauranteId)
    const supabase = await criarClienteDoServidor()

    const { data: item, error } = await supabase
      .from("cart_items")
      .insert({
        cart_id: carrinhoId,
        product_id: dados.produtoId,
        quantity: Math.max(1, Math.min(99, dados.quantidade)),
        notes: dados.observacao?.trim() || null,
      })
      .select("id")
      .single()

    if (error || !item) return { ok: false, erro: error?.message ?? "Não consegui adicionar." }

    if (dados.adicionais.length > 0) {
      const { error: erroDosAdicionais } = await supabase.from("cart_item_addons").insert(
        dados.adicionais.map((addon_id) => ({ cart_item_id: item.id, addon_id })),
      )
      if (erroDosAdicionais) return { ok: false, erro: erroDosAdicionais.message }
    }

    recarregar()
    return { ok: true, id: item.id }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : "Não consegui adicionar." }
  }
}

export async function mudarQuantidade(
  itemId: string,
  quantidade: number,
): Promise<ResultadoDaAcao> {
  await exigirUsuario("/carrinho")
  const supabase = await criarClienteDoServidor()

  const { error } =
    quantidade <= 0
      ? await supabase.from("cart_items").delete().eq("id", itemId)
      : await supabase
          .from("cart_items")
          .update({ quantity: Math.min(99, quantidade) })
          .eq("id", itemId)

  if (error) return { ok: false, erro: error.message }

  recarregar()
  return { ok: true }
}

export async function esvaziarCarrinho(): Promise<ResultadoDaAcao> {
  const contexto = await exigirUsuario("/carrinho")
  const supabase = await criarClienteDoServidor()

  const { error } = await supabase.from("carts").delete().eq("user_id", contexto.id)
  if (error) return { ok: false, erro: error.message }

  recarregar()
  return { ok: true }
}

/**
 * Fecha o pedido.
 *
 * Manda o que a pessoa escolheu — jamais quanto custa. `fechar_pedido`
 * recalcula tudo a partir do cardapio daquele instante, e e o unico caminho
 * pelo qual um cliente cria pedido: a politica de INSERT em `orders` nao
 * aceita mais nada vindo dele.
 */
export async function fecharPedido(dados: {
  tipo: Enums<"fulfillment_type">
  forma: Enums<"payment_method">
  enderecoId?: string
  cupom?: string
  observacao?: string
  trocoPara?: string
}): Promise<ResultadoDaAcao> {
  const contexto = await exigirUsuario("/carrinho")
  const supabase = await criarClienteDoServidor()

  const { data: carrinho } = await supabase
    .from("carts")
    .select("id")
    .eq("user_id", contexto.id)
    .maybeSingle()

  if (!carrinho) return { ok: false, erro: "Seu carrinho está vazio." }

  // Pix e cartão de crédito são as formas que um gateway processa; o resto se
  // paga na porta. Enquanto o provedor é simulado, a tela prefere as de porta
  // — um pedido em "aguardando pagamento" nunca chegaria ao balcão.
  const noApp = dados.forma === "pix" || dados.forma === "credit_card"

  const troco = dados.trocoPara?.trim()
    ? Math.round(Number(dados.trocoPara.replace(/[^\d,.-]/g, "").replace(",", ".")) * 100)
    : null

  const { data: id, error } = await supabase.rpc("fechar_pedido", {
    p_cart_id: carrinho.id,
    p_fulfillment: dados.tipo,
    p_payment_method: dados.forma,
    p_payment_timing: noApp ? "online" : "on_delivery",
    p_address_id: dados.tipo === "delivery" ? (dados.enderecoId ?? undefined) : undefined,
    p_change_for_cents:
      dados.forma === "cash" && troco !== null ? troco : undefined,
    p_coupon_code: dados.cupom?.trim() || undefined,
    p_notes: dados.observacao?.trim() || undefined,
  })

  if (error) return { ok: false, erro: error.message }

  recarregar()
  revalidatePath("/pedidos")
  return { ok: true, id: id as string }
}

/**
 * Quanto um cupom abate, antes de fechar.
 *
 * A conta e do banco: `simular_cupom` usa a MESMA funcao de desconto que
 * `fechar_pedido` usa na hora de cobrar. Se a tela fizesse a propria conta, as
 * duas divergiriam no dia em que alguem mexesse numa delas.
 */
export async function simularCupom(dados: {
  codigo: string
  restauranteId: string
  subtotalCentavos: number
  taxaCentavos: number
}): Promise<{ ok: true; descontoCentavos: number } | { ok: false; erro: string }> {
  await exigirUsuario("/carrinho")
  const supabase = await criarClienteDoServidor()

  const { data, error } = await supabase.rpc("simular_cupom", {
    p_code: dados.codigo.trim(),
    p_restaurant: dados.restauranteId,
    p_subtotal_cents: dados.subtotalCentavos,
    p_delivery_fee_cents: dados.taxaCentavos,
  })

  // As mensagens vem do banco ja em portugues — "Este cupom venceu", "Voce ja
  // usou este cupom" — e chegam a tela como estao.
  if (error) return { ok: false, erro: error.message }

  const resposta = data as { discount_cents?: number } | null
  return { ok: true, descontoCentavos: Number(resposta?.discount_cents ?? 0) }
}

/** Quanto o carrinho tem, para o contador do cabeçalho. */
export async function contarItensDoCarrinho(): Promise<number> {
  const contexto = await obterContexto()
  if (!contexto) return 0

  const supabase = await criarClienteDoServidor()
  const { data } = await supabase
    .from("carts")
    .select("cart_items(quantity)")
    .eq("user_id", contexto.id)
    .maybeSingle()

  return (data?.cart_items ?? []).reduce((soma, i) => soma + i.quantity, 0)
}

File: celular/lib/dados/carrinho.dart
/// O carrinho — que mora no servidor, não no celular.
///
/// A tabela `carts` existe justamente para isso: quem monta o pedido no ônibus
/// e fecha em casa encontra o carrinho como deixou. E, mais importante, é do
/// carrinho do servidor que `fechar_pedido` lê para recalcular os preços. Um
/// carrinho só local obrigaria o aplicativo a mandar o que custa, que é
/// exatamente o que o banco não aceita.
///
/// O preço que esta classe soma serve para *mostrar*. Quem cobra é o banco.
library;

import 'package:flutter/foundation.dart';

import '../modelos/modelos.dart';
import 'supabase.dart';

class AdicionalEscolhido {
  const AdicionalEscolhido({required this.adicional, this.quantidade = 1});
  final Adicional adicional;
  final int quantidade;

  int get totalCentavos => adicional.precoCentavos * quantidade;
}

class ItemDoCarrinho {
  const ItemDoCarrinho({
    required this.id,
    required this.produto,
    required this.quantidade,
    required this.adicionais,
    this.observacao,
  });

  final String id;
  final Produto produto;
  final int quantidade;
  final List<AdicionalEscolhido> adicionais;
  final String? observacao;

  int get adicionaisCentavos =>
      adicionais.fold(0, (s, a) => s + a.totalCentavos);

  int get totalCentavos =>
      (produto.precoQueVale + adicionaisCentavos) * quantidade;
}

class Carrinho extends ChangeNotifier {
  Carrinho._();
  static final Carrinho instancia = Carrinho._();

  String? _id;
  Restaurante? _restaurante;
  List<ItemDoCarrinho> _itens = const [];
  bool _carregando = false;

  String? get id => _id;
  Restaurante? get restaurante => _restaurante;
  List<ItemDoCarrinho> get itens => _itens;
  bool get carregando => _carregando;
  bool get vazio => _itens.isEmpty;

  int get quantidadeDeItens => _itens.fold(0, (s, i) => s + i.quantidade);
  int get subtotalCentavos => _itens.fold(0, (s, i) => s + i.totalCentavos);

  /// Quanto falta para o pedido mínimo. Zero quando já dá.
  int faltaParaOMinimo() {
    final minimo = _restaurante?.pedidoMinimoCentavos ?? 0;
    final falta = minimo - subtotalCentavos;
    return falta > 0 ? falta : 0;
  }

  /// Lê o carrinho aberto da pessoa, se houver.
  ///
  /// O índice `carts_open_per_restaurant_idx` garante um carrinho por
  /// estabelecimento; na prática o aplicativo trabalha com um de cada vez, e
  /// trocar de restaurante troca o carrinho.
  Future<void> carregar() async {
    if (usuarioId == null) {
      _id = null;
      _restaurante = null;
      _itens = const [];
      notifyListeners();
      return;
    }

    _carregando = true;
    notifyListeners();

    try {
      final linhas = await executar(() => banco
          .from('carts')
          .select('id, restaurant_id, restaurants(*)')
          .order('updated_at', ascending: false)
          .limit(1));

      if (linhas.isEmpty) {
        _id = null;
        _restaurante = null;
        _itens = const [];
        return;
      }

      _id = linhas.first['id'] as String;
      _restaurante =
          Restaurante.deMapa(linhas.first['restaurants'] as Map<String, dynamic>);
      await _lerItens();
    } finally {
      _carregando = false;
      notifyListeners();
    }
  }

  Future<void> _lerItens() async {
    final id = _id;
    if (id == null) {
      _itens = const [];
      return;
    }

    final linhas = await executar(() => banco
        .from('cart_items')
        .select('*, products(*), cart_item_addons(quantity, addons(*))')
        .eq('cart_id', id)
        .order('created_at'));

    _itens = linhas.map((l) {
      final adicionais = ((l['cart_item_addons'] as List?) ?? const [])
          .cast<Map<String, dynamic>>()
          .map((a) => AdicionalEscolhido(
                adicional: Adicional.deMapa(a['addons'] as Map<String, dynamic>),
                quantidade: (a['quantity'] as int?) ?? 1,
              ))
          .toList();

      return ItemDoCarrinho(
        id: l['id'] as String,
        produto: Produto.deMapa(l['products'] as Map<String, dynamic>),
        quantidade: l['quantity'] as int,
        observacao: l['notes'] as String?,
        adicionais: adicionais,
      );
    }).toList();
  }

  /// Abre (ou reabre) o carrinho de um estabelecimento.
  ///
  /// Trocar de restaurante apaga o carrinho anterior. É a regra do schema — um
  /// pedido pertence a um estabelecimento — e quem confirma isso com o cliente
  /// é a tela, antes de chamar aqui.
  Future<void> abrirEm(Restaurante restaurante) async {
    if (usuarioId == null) {
      throw ErroDeDados('Entre na sua conta para montar o pedido.');
    }
    if (_id != null && _restaurante?.id == restaurante.id) return;

    await executar(() async {
      if (_id != null) await banco.from('carts').delete().eq('id', _id!);

      final linha = await banco
          .from('carts')
          .insert({'user_id': usuarioId, 'restaurant_id': restaurante.id})
          .select('id')
          .single();

      _id = linha['id'] as String;
    });

    _restaurante = restaurante;
    _itens = const [];
    notifyListeners();
  }

  /// O carrinho aberto é de outro estabelecimento?
  bool ehDeOutro(String restauranteId) =>
      _id != null && !vazio && _restaurante?.id != restauranteId;

  Future<void> adicionar({
    required Restaurante restaurante,
    required Produto produto,
    required int quantidade,
    List<AdicionalEscolhido> adicionais = const [],
    String? observacao,
  }) async {
    await abrirEm(restaurante);

    await executar(() async {
      final item = await banco
          .from('cart_items')
          .insert({
            'cart_id': _id,
            'product_id': produto.id,
            'quantity': quantidade,
            'notes': (observacao?.isEmpty ?? true) ? null : observacao,
          })
          .select('id')
          .single();

      if (adicionais.isNotEmpty) {
        await banco.from('cart_item_addons').insert([
          for (final a in adicionais)
            {
              'cart_item_id': item['id'],
              'addon_id': a.adicional.id,
              'quantity': a.quantidade,
            }
        ]);
      }
    });

    await _lerItens();
    notifyListeners();
  }

  Future<void> alterarQuantidade(String itemId, int quantidade) async {
    if (quantidade <= 0) return remover(itemId);

    await executar(() => banco
        .from('cart_items')
        .update({'quantity': quantidade}).eq('id', itemId));

    await _lerItens();
    notifyListeners();
  }

  Future<void> remover(String itemId) async {
    await executar(() => banco.from('cart_items').delete().eq('id', itemId));
    await _lerItens();
    if (_itens.isEmpty) await esvaziar();
    notifyListeners();
  }

  Future<void> esvaziar() async {
    final id = _id;
    if (id == null) return;
    await executar(() => banco.from('carts').delete().eq('id', id));
    _id = null;
    _restaurante = null;
    _itens = const [];
    notifyListeners();
  }

  /// Esquece o carrinho sem apagar nada — usado depois do fechamento, quando o
  /// próprio `fechar_pedido` já removeu a linha no banco.
  void esquecer() {
    _id = null;
    _restaurante = null;
    _itens = const [];
    notifyListeners();
  }
}
